app/models/concerns/orchestration/external_ipam.rb

```
@@ -0,0 +1,113 @@
+module Orchestration::ExternalIPAM
+  extend ActiveSupport::Concern
+  include Orchestration::Common
+  include SubnetsHelper
+
+  included do
+    after_validation :queue_external_ipam
+    before_destroy :initialize_ip_subnet
+    before_destroy :queue_external_ipam_destroy
+  end
+
+  def generate_external_ipam_task_id(action, interface = self)
+    id = [interface.mac, interface.ip, interface.identifier, interface.id].find {|x| x&.present?}
+    "external_ipam_#{action}_#{id}"
+  end
+
+  protected
+
+  def set_external_ip
+    initialize_ip_subnet
+    if ip_is_available?
+      response = @subnet.external_ipam_proxy.add_ip_to_subnet(@ip, @subnet)
+      success?(response, 'Address created')
+    else
+      self.errors.add ipv6? ? :ip6 : :ip, _('This IP address has already been reserved in External IPAM')
+      self.errors.add :interfaces, _('Some interfaces are invalid')
+    end
+  end
+
+  # Empty method for rollbacks. We don't want to delete IP's from IPAM when there
+  # is a conflict(i.e. IP address already taken)
+  def del_external_ip
+  end
+
+  def remove_external_ip
+    initialize_ip_subnet
+    response = @subnet.external_ipam_proxy.delete_ip_from_subnet(@ip, @subnet)
+    success?(response, 'Address deleted')
+  end
+
+  def requires_update?
+    return false if new_record?
+    old.ip != self.ip
+  end
+
+  def requires_update6?
+    return false if new_record?
+    old.ip6 != self.ip6
+  end
+
+  def requires_delete?
+    old = Nic::Base.find(id)
+    !old.ip.nil? && !old.subnet_id.nil?
+  end
+
+  def requires_delete6?
+    old = Nic::Base.find(id)
+    !old.ip6.nil? && !old.subnet6_id.nil?
+  end
+
+  def ip_is_available?
+    response = @subnet.external_ipam_proxy.ip_exists(@ip, @subnet)
+    success?(response, 'No addresses found')
+  end
+
+  def success?(response, message)
+    (response['message'] && response['message'] == message) ? true : false
+  end
```

One problem with this is when remote system does not return the expected success message, false is returned but the actual error is swallowed and never shown to the user. I suggest to raise an exception here and put the message into the exception message. Suggested exception is Foreman::Exception which also generates an ERF code: https://community.theforeman.org/t/foreman-error-codes/1064